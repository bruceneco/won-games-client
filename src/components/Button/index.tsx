import * as S from './styles'
import { ButtonHTMLAttributes } from 'react'

export type ButtonProps = {
  children?: React.ReactNode
  size?: 'small' | 'medium' | 'large'
  fullWidth?: boolean
  icon?: React.ReactNode
} & ButtonHTMLAttributes<HTMLButtonElement>

const Button = ({
  children,
  size = 'medium',
  fullWidth = false,
  icon,
  ...props
}: ButtonProps) => (
  <S.Wrapper size={size} fullWidth={fullWidth} hasIcon={!!icon} {...props}>
    {!!icon && icon}
    {!!children && <span>{children}</span>}
  </S.Wrapper>
)

export default Button
