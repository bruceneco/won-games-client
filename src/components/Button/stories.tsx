import { Story, Meta } from '@storybook/react'
import Button, { ButtonProps } from '.'
import { AddShoppingCart } from '@styled-icons/material-outlined'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    children: {
      type: 'string'
    },
    icon: {
      control: { disable: true }
    }
  }
} as Meta

export const Default: Story<ButtonProps> = (args) => <Button {...args} />

Default.args = {
  children: 'Buy now'
}

export const WithIcon: Story<ButtonProps> = (args) => <Button {...args} />
WithIcon.args = {
  size: 'small',
  children: 'Buy now',
  icon: <AddShoppingCart />
}
