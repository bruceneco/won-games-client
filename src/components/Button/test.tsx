import { screen } from '@testing-library/react'

import Button from '.'
import { renderWithTheme } from '../../utils/tests/helpers'
import { AddShoppingCart } from '@styled-icons/material-outlined'

describe('<Button />', () => {
  it('should render the medium size button by default', function () {
    const { container } = renderWithTheme(<Button>Buy now</Button>)

    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
      height: '4rem',
      padding: '0.8rem 3.2rem',
      fontSize: '1.4rem'
    })

    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render the small size', function () {
    renderWithTheme(<Button size="small">Buy now</Button>)

    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
      height: '3rem',
      fontSize: '1.2rem'
    })
  })
  it('should render the large size', function () {
    renderWithTheme(<Button size="large">Buy now</Button>)

    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
      height: '5rem',
      fontSize: '1.6rem',
      padding: '0.8rem 4.8rem'
    })
  })
  it('should render the full width version', function () {
    renderWithTheme(<Button fullWidth>Buy now</Button>)

    expect(screen.getByRole('button', { name: /buy now/i })).toHaveStyle({
      width: '100%'
    })
  })
  it('should render an icon version', function () {
    renderWithTheme(
      <Button icon={<AddShoppingCart data-testid="icon" />}>Buy now</Button>
    )

    expect(screen.getByText(/buy now/i)).toBeInTheDocument()
    expect(screen.getByTestId('icon')).toBeInTheDocument()
  })
})
