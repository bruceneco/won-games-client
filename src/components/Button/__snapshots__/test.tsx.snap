// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render the medium size button by default 1`] = `
.c0 {
  background: linear-gradient(180deg,#ff5f5f 0%,#f062c0 100%);
  color: #FAFAFA;
  border: 0;
  border-radius: 0.4rem;
  padding: 0.8rem;
  cursor: pointer;
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  height: 4rem;
  font-size: 1.4rem;
  padding: 0.8rem 3.2rem;
}

.c0:hover {
  -webkit-filter: brightness(0.95);
  filter: brightness(0.95);
}

<button
  class="c0"
>
  <span>
    Buy now
  </span>
</button>
`
